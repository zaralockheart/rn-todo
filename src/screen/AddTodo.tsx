import React from 'react'
import {SafeAreaView, StyleSheet, Text, TouchableOpacity, View} from 'react-native'
import {ErrorOther, Field, InjectedFormProps, reduxForm} from 'redux-form'
import _ from 'lodash'
import {useNavigation, useNavigationParam} from 'react-navigation-hooks'

import {Input, ListSelector} from '../components/form'
import {useDispatch} from 'react-redux'
import {addTodo, editTodo} from '../redux'
import {isNotEmpty} from '../util/main'

export const days = ['MON', 'TUE', 'WED', 'THU', 'FRI', 'SAT', 'SUN']

interface TodoFormProps {
  title: string
  subtitle: string
  days: { [key: number]: string }
}

const AddTodo = (props: InjectedFormProps<TodoFormProps>) => {
  const {goBack} = useNavigation()

  const dispatch = useDispatch()

  const id = useNavigationParam('id')

  const save = (obj: TodoFormProps) => {
    if (id) {
      dispatch(
        editTodo({
          id,
          days: obj.days,
          title: obj.title,
          subtitle: obj.subtitle
        })
      )
    } else {
      dispatch(
        addTodo({
          days: obj.days,
          title: obj.title,
          subtitle: obj.subtitle
        })
      )
    }

    goBack()
  }

  return (
    <SafeAreaView style={{flex: 1}}>
      <View style={styles.parent}>
        <Text>Title</Text>
        <Field name="title" component={Input} placeholder={'What is is about?'}/>
        <Text>Subtitle</Text>
        <Field
          name="subtitle"
          component={Input}
          placeholder={'Any more details?'}
        />
        <Field
          name="days"
          component={ListSelector}
          items={days.map((day, index) => ({
            id: index,
            text: day
          }))}
        />
        <View style={{flex: 1}}/>
        <TouchableOpacity
          style={styles.save}
          disabled={props.invalid}
          onPress={props.handleSubmit(save)}>
          <Text>Save</Text>
        </TouchableOpacity>
      </View>
    </SafeAreaView>
  )
}

const AddTodoForm = reduxForm<TodoFormProps>({
  form: 'TODO_FORM',
  enableReinitialize: true,
  validate(
    values: TodoFormProps
  ): { [P in keyof TodoFormProps]?: React.ReactElement | string } &
    ErrorOther<string> {
    const errors: Partial<{ [x in keyof TodoFormProps]: string }> = {
      title: undefined,
      days: undefined
    }

    errors.title = isNotEmpty(values.title, 'Title is required')
    errors.days = _.isEmpty(values.days)
      ? 'Required at least on day'
      : undefined

    return errors
  }
})(AddTodo)

export default () => {
  const subtitle = useNavigationParam('subtitle')
  const title = useNavigationParam('title')
  const days = useNavigationParam('days')

  return (
    <AddTodoForm
      initialValues={{
        title,
        subtitle,
        days
      }}
    />
  )
}

const styles = StyleSheet.create({
  parent: {
    flex: 1,
    padding: 10,
  },
  save: {
    backgroundColor: 'brown',
    padding: 20,
    alignItems: 'center',
    borderRadius: 10,
  }
})
