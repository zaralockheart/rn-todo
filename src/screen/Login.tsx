import React from 'react'
import {ErrorOther, Field, InjectedFormProps, reduxForm} from 'redux-form'
import {SafeAreaView, Text, TouchableOpacity} from "react-native"
import TextInput from '../components/form'
import {isNotEmpty} from "../util/string_util"
import {useNavigation} from "react-navigation-hooks"

// import {useNavigation} from 'react-navigation-hooks'

interface LoginProps {
	email: string
	password: string
}

const LoginForm = (props: InjectedFormProps<LoginProps>) => {
	const {handleSubmit} = props

	const {navigate} = useNavigation();

	const validateLogin = ({email, password}: LoginProps) => {
		navigate('Home');
	}

	return (
		<SafeAreaView>
			<Text>Email:</Text>
			<Field name="email" component={TextInput}/>
			<Field name="password" component={TextInput}/>
			<TouchableOpacity onPress={handleSubmit(validateLogin)}>
				<Text>Submit</Text>
			</TouchableOpacity>
		</SafeAreaView>
	)
}

export default reduxForm<LoginProps>({
	form: 'LOGIN_FORM',
	validate(values: LoginProps, props: {} & InjectedFormProps<LoginProps, {}, string>): { [P in keyof LoginProps]?: React.ReactElement | string } & ErrorOther<string> {
		const errors: Partial<LoginProps> = {
			email: undefined,
			password: undefined,
		};
		errors.email = isNotEmpty(values.email)

		errors.password = isNotEmpty(values.password)

		return errors;
	}
})(LoginForm)
