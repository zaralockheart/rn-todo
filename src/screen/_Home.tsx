import React from "react"
import {StyleSheet, Text, TouchableOpacity, View} from "react-native"

import {TodoObject} from "../redux"
import {days} from "./AddTodo"

interface TodoCardProps {
	onPressRemove: (id: string) => void
	onPressEdit: (id: string) => void
}

export const TodoCard = (args: TodoObject & TodoCardProps) => {
	return (
		<TouchableOpacity style={styles.parent} onPress={() => args.onPressEdit(args.id)}>
			<View style={styles.row}>
				<Text>{args.title}</Text>
				<TouchableOpacity
					style={styles.removeButton}
					onPress={() => args.onPressRemove(args.id)}/>
			</View>
			<View style={[styles.row]}>
				{days.map((day, index) => (
					<View
						key={index}
						style={[styles.round, !!args.days[index] && styles.buttonSelected]}>
						<Text>{day}</Text>
					</View>
				))}
			</View>
		</TouchableOpacity>
	)
}

const styles = StyleSheet.create({
	parent: {
		padding: 10
	},
	row: {
		flexDirection: 'row',
		justifyContent: 'space-between',
	},
	removeButton: {
		height: 20,
		width: 20,
		backgroundColor: 'red'
	},
	buttonSelected: {
		backgroundColor: 'blue',
	},
	round: {
		padding: 5,
		height: 50,
		width: 50,
		borderRadius: 100,
		justifyContent: 'center',
		alignItems: 'center',
	}
})
