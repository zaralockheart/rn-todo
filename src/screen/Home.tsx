import React from 'react'
import {
  FlatList,
  SafeAreaView,
  StyleSheet,
  Text,
  TouchableOpacity,
  View
} from 'react-native'
import {useNavigation} from 'react-navigation-hooks'
import {useDispatch, useSelector} from 'react-redux'
import {removeTodo, State} from '../redux'
import {TodoCard} from './_Home'

export default () => {
  const {navigate} = useNavigation()

  const todos = useSelector((state: State) => state.todo)

  const dispatch = useDispatch()

  return (
    <React.Fragment>
      <SafeAreaView>
        <View style={style.appbar}>
          <Text>Home</Text>
          <TouchableOpacity
            style={style.addButton}
            onPress={() => navigate('AddTodo')}
          />
        </View>
        <FlatList
          style={{height: '100%'}}
          data={todos.ids}
          renderItem={({item}) => (
            <TodoCard
              onPressEdit={id => {
                navigate('AddTodo', {
                  ...todos.todo[id]
                })
              }}
              onPressRemove={id => dispatch(removeTodo({id}))}
              {...todos.todo[item]}
            />
          )}
          keyExtractor={item => item}
          extraData={todos}
        />
      </SafeAreaView>
    </React.Fragment>
  )
}

const style = StyleSheet.create({
  appbar: {
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  addButton: {
    height: 20,
    width: 20,
    backgroundColor: 'red'
  }
})
