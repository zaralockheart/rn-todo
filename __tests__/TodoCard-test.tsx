import {shallow} from "enzyme";
import React from "react";
import {TodoCard} from "../src/screen/_Home";
import renderer from "react-test-renderer";

describe('Add Todo Form', () => {
    it('renders correctly', () => {
        const component = renderer.create(
            <TodoCard
                days={
                    {1: "Monday"}
                }
                id={"1234"}
                title={'title'}
                subtitle={'subtitle'}
                onPressEdit={jest.fn}
                onPressRemove={jest.fn}
            />
        )

        expect(component.toJSON()).toMatchSnapshot()
    })
})
