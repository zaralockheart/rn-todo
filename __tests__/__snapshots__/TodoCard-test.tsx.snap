// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Todo Form renders correctly 1`] = `
<View
  accessible={true}
  clickable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
      "padding": 10,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text>
      title
    </Text>
    <View
      accessible={true}
      clickable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "red",
          "height": 20,
          "opacity": 1,
          "width": 20,
        }
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 100,
            "height": 50,
            "justifyContent": "center",
            "padding": 5,
            "width": 50,
          },
          false,
        ]
      }
    >
      <Text>
        MON
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 100,
            "height": 50,
            "justifyContent": "center",
            "padding": 5,
            "width": 50,
          },
          Object {
            "backgroundColor": "blue",
          },
        ]
      }
    >
      <Text>
        TUE
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 100,
            "height": 50,
            "justifyContent": "center",
            "padding": 5,
            "width": 50,
          },
          false,
        ]
      }
    >
      <Text>
        WED
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 100,
            "height": 50,
            "justifyContent": "center",
            "padding": 5,
            "width": 50,
          },
          false,
        ]
      }
    >
      <Text>
        THU
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 100,
            "height": 50,
            "justifyContent": "center",
            "padding": 5,
            "width": 50,
          },
          false,
        ]
      }
    >
      <Text>
        FRI
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 100,
            "height": 50,
            "justifyContent": "center",
            "padding": 5,
            "width": 50,
          },
          false,
        ]
      }
    >
      <Text>
        SAT
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 100,
            "height": 50,
            "justifyContent": "center",
            "padding": 5,
            "width": 50,
          },
          false,
        ]
      }
    >
      <Text>
        SUN
      </Text>
    </View>
  </View>
</View>
`;
